### ISSUE TITLE
tkiptun-ng improvment - Workaround for unreported MIC failures.

### ISSUE THREAD

**Issue by aircrack-ng** [aircrack-ng/aircrack-ng#582]
**Reported by misterx on 16 Feb 2009 23:17 UTC**

Source thread: http://tinyshell.be/aircrackng/forum/index.php?topic=4476.msg27028#msg27028

From my experience with tkiptun-ng, there is a common problem when running the chopchop attack, that the MIC failure report may not be detected. This will make the program continue sending packets and if a failure report actually was sent by the AP, it will cause the AP to activate its contermeasures and shut down for 60 seconds. This will make the tkiptun-ng attack fail.

To cope with this, I re-wrote the following section (line 2487 - 2494) to sleep for 60 seconds when not detecting a MIC failure after trying all permutations for the chopchop-byte. From my experience, the attack now has a much higher success rate than before:

```
Index: src/tkiptun-ng.c
===================================================================
--- src/tkiptun-ng.c    (revision 1414)
+++ src/tkiptun-ng.c    (working copy)
@@ -2485,11 +2485,19 @@
         /* update the status line */

         if( ticks[> (RTC_RESOLUTION/10) )
-        {
-            ticks[1](1]) = 0;
-            printf( "\rSent %3ld packets, current guess: %02X...\33[                   nb_pkt_sent, guess );
-            fflush( stdout );
+               {
+                       ticks[1](K",
-) = 0;
+                       printf( "\rSent %3ld packets, current guess: %02X...\33[                                      nb_pkt_sent, guess );
+                       fflush( stdout );
+                       if( (nb_pkt_sent > 0) && (nb_pkt_sent % 256 == 0))
+                       {
+                               printf( "\rLooks like mic failure report was not detected."
+                                               "Waiting 60 seconds before trying again to avoid "
+                                               "the AP shutting down.");
+                               fflush( stdout );
+                               sleep(60);
+                       }
         }

 /*        if( data_end < 47 && ticks[3](K",
+) > 8 * ( ticks[- ticks[3](0]) ) /

```

**Comment by aircrack-ng** [comment]
**Comment by misterx on 16 Feb 2009 23:17 UTC**

When chopchop-ing a byte, it will try all permutations of that byte, in total 256 permutations. By nature, one of these should result in a MIC failure coming from the AP. If one doesn't receive a MIC failure after the first 256 permutations, it may be because you simply didn't detect it. If this is the case, and you continue sending suggestions for the byte to the AP, the AP will eventually shut down and re-key the next time you hit the correct byte. This will mess up the attack and you would have to start over again.

Instead, by waiting 60 seconds after those first 256 permutations, you can avoid the AP from getting suspicious about what's going on.

**Comment by aircrack-ng** [comment]
**Comment by misterx on 16 Feb 2009 23:17 UTC**

Another improvment thanks to buggy cisco phones:

```
<jwright> There are a lot of Cisco products such as the 7920 phone that
only support TKIP and have buggy software implementations, causing them
to frequently send MIC failure messages.
<jwright> Cisco recommends that users reduce the hold time for triggering
countermeasures as a workaround to 0-3 seconds.
<jwright> Allowing the tkiptun-ng attack to be significantly accelerated,
recovering a new plaintext byte at 3-second intervals.
```

**Comment by aircrack-ng** [comment]
**Comment by hirte on 16 Feb 2009 23:17 UTC**

fixed in aircrack-ng/aircrack-ng@55448b0c09f75b0f9b477378c967cc96cca2ee5b, thanks for issuing the improvement
